python/tank/platform/environment.py

```diff
@@ -878,3 +898,284 @@ def copy_apps(self, src_engine_name, dst_engine_name):
+        """
+
+        self._dump_settings(self.SPARSE, file, debug)
+
+    def dump_full(self, file, debug=False):
+        """
+        Dump a full copy of this environment's settings to a file.
+
+        Any setting not specified in the environment will be dumped with the
+        default value.
+
+        NOTE: The calling code is responsible for closing the file handle.
+
+        :param file: An open file object to write to.
+        :param debug: If True, include debug comments on lines using a default
+            value.
+        """
+
+        self._dump_settings(self.FULL, file, debug)
+
+    def _update_settings_sparse(self, schema, settings, engine_name=None,
+        manifest_file=None, debug=False):
+        """
+        Given a schema and settings, make sure that the settings are sparse.
+
+        Iterate over the schema key's default values and remove any settings
+        that use the default.
+
+        :param schema: A schema defining types and defaults for settings.
+        :param settings: A dict of settings to sparsify.
+        :param engine_name: The name of the current engine
+        :param manifest_file: The path to the manifest file if known.
+        :param debug: If True, include debug comments on lines using a
+            non-default value.
+
+        :returns: bool - True if the settings were modified, False otherwise.
+        """
+
+        modified = False
+
+        # check each key defined in the schema
+        for setting_name in schema.keys():
+
+            # identify the portion of the manifest that defines the setting
+            setting_schema = schema[setting_name]
+
+            # if defined, check the settings against the default value
+            if setting_name in settings:
+                setting_value = settings[setting_name]
+                schema_default = resolve_default_value(
+                    setting_schema, engine_name=engine_name)
+
+                if setting_value == schema_default:
+                    # the setting value matches the schema default. remove the
+                    # setting.
+                    del settings[setting_name]
+                    modified = True
+
+                elif debug:
+                    # Add a comment to identify default values in the dumped
+                    # file to help with debugging. The comment shows the
+                    # default value and the path to the manifest where the
+                    # default value is defined.
+                    if hasattr(settings, 'yaml_add_eol_comment'):
+                        settings.yaml_add_eol_comment(
+                            "Differs from %s default value (%s) in manifest %s"
+                                % (engine_name or "",
+                                   schema_default or '""',
+                                   manifest_file or ""
+                                ),
+                            setting_name,
+                            column=80
```

Comment:
Here's a screenshot so you can see what the comments look like. I tried a few values, and this seemed to be pretty reasonable. Let me know if you think it should be changed. 

![screen shot 2016-04-11 at 8 50 34 am](https://cloud.githubusercontent.com/assets/2604646/14427623/a6ae8aa2-ffc3-11e5-8bed-38f03eec8d72.png)